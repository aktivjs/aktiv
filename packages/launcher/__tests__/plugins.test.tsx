import Launcher, { Outlet, useNavigate } from '../src';
import userEvent from '@testing-library/user-event';
import { act, screen } from '@testing-library/react';
import { ReactNode } from 'react';
// eslint-disable-next-line init-declarations
let node: HTMLDivElement;

beforeEach(() => {
    node = document.createElement('div');

    node.id = 'root';
    node.setAttribute('data-testid', 'root');
    document.body.appendChild(node);
});

afterEach(() => {
    node.remove();
    window.history.pushState({}, '', '/');
});
describe('plugin', () => {
    describe('plugin with router', () => {
        it('should inner wrap the route', () => {
            /* eslint-disable */
            const spy = jest.fn(() => {});
            /* eslint-enable */

            const opt = {
                a: 1,
                b: 2,
            };
            const plugin = {
                name: 'testPlugin',
                inner: (children: ReactNode, receiveOpt: any) => {
                    expect(receiveOpt).toEqual(opt);
                    spy();

                    return (
                        <div>
                            innerRouter
                            <div>{children}</div>
                        </div>
                    );
                },
            };
            const Home = () => {
                const navigate = useNavigate();

                return (
                    <div>
                        <button
                            onClick={() => {
                                navigate('children');
                            }}
                        >
                            go children
                        </button>
                        home
                        <Outlet />
                    </div>
                );
            };
            const Children = () => {
                return <div>children</div>;
            };
            const app = new Launcher({
                hash: false,
                routes: [
                    {
                        path: '/',
                        component: Home,
                        children: [
                            {
                                path: 'children',
                                component: Children,
                            },
                        ],
                    },
                ],
            });

            app.use(plugin, opt);
            act(() => {
                app.start();
            });
            expect(document.body).toMatchSnapshot();
            act(() => {
                userEvent.click(screen.getByText('go children'));
            });
            expect(document.body).toMatchSnapshot();
            expect(spy).toHaveBeenCalledTimes(2);
        });
    });
});
