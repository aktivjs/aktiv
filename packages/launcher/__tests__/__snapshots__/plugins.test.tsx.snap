// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin plugin with router should inner wrap the route 1`] = `
<body>
  <div
    data-testid="root"
    id="root"
  >
    <div>
      innerRouter
      <div>
        <div>
          <button>
            go children
          </button>
          home
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`plugin plugin with router should inner wrap the route 2`] = `
<body>
  <div
    data-testid="root"
    id="root"
  >
    <div>
      innerRouter
      <div>
        <div>
          <button>
            go children
          </button>
          home
          <div>
            innerRouter
            <div>
              <div>
                children
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
